include/boost/compute/algorithm/detail/copy_on_device.hpp

```diff
@@ -27,93 +27,96 @@ namespace boost {
+                                size_t count,
+                                command_queue &queue)
 {
-public:
-    copy_kernel(const device &device)
-        : meta_kernel("copy")
-    {
-        m_count = 0;
-
-        typedef typename std::iterator_traits<InputIterator>::value_type input_type;
-
-        boost::shared_ptr<parameter_cache> parameters =
-            detail::parameter_cache::get_global_cache(device);
+    typedef typename std::iterator_traits<InputIterator>::value_type input_type;
+
+    const device& device = queue.get_device();
+    boost::shared_ptr<parameter_cache> parameters =
+        detail::parameter_cache::get_global_cache(device);
+    std::string cache_key =
+        "__boost_copy_kernel_" + boost::lexical_cast<std::string>(sizeof(input_type));
+
+    uint_ vpt = parameters->get(cache_key, "vpt", 4);
+    uint_ tpb = parameters->get(cache_key, "tpb", 128);
+
+    meta_kernel k("copy");
+    k <<
+        "uint index = get_local_id(0) + " <<
+            "(" << vpt * tpb << " * get_group_id(0));\n" <<
+        "for(uint i = 0; i < " << vpt << "; i++){\n" <<
+        "    if(index < count){\n" <<
+                result[k.var<uint_>("index")] << '=' <<
+                    first[k.var<uint_>("index")] << ";\n" <<
+        "       index += " << tpb << ";\n"
+        "    }\n"
+        "}\n";
+
+    k.add_set_arg<const uint_>("count", static_cast<uint_>(count));
+    size_t global_work_size = calculate_work_size(count, vpt, tpb);
+    return k.exec_1d(queue, 0, global_work_size, tpb);
+}
 
-        std::string cache_key =
-            "__boost_copy_kernel_" + boost::lexical_cast<std::string>(sizeof(input_type));
+template<class InputIterator, class OutputIterator>
+inline event dispatch_copy_on_device(InputIterator first,
+                                     InputIterator last,
+                                     OutputIterator result,
+                                     command_queue &queue)
+{
+    const size_t count = detail::iterator_range_size(first, last);
 
-        m_vpt = parameters->get(cache_key, "vpt", 4);
-        m_tpb = parameters->get(cache_key, "tpb", 128);
+    if(count == 0){
+        // nothing to do
+        return event();
     }
 
-    void set_range(InputIterator first,
-                   InputIterator last,
-                   OutputIterator result)
-    {
-        m_count_arg = add_arg<uint_>("count");
-
-        *this <<
-            "uint index = get_local_id(0) + " <<
-               "(" << m_vpt * m_tpb << " * get_group_id(0));\n" <<
-            "for(uint i = 0; i < " << m_vpt << "; i++){\n" <<
-            "    if(index < count){\n" <<
-                     result[expr<uint_>("index")] << '=' <<
-                         first[expr<uint_>("index")] << ";\n" <<
-            "        index += " << m_tpb << ";\n"
-            "    }\n"
-            "}\n";
-
-        m_count = detail::iterator_range_size(first, last);
+#ifndef __APPLE__
```

Comment:
> `device.platform().name() == "Apple"`

I can do this. I was just worried about using those kind of conditions in code. `__APPLE__` seemed more reliable.

> Either way, could you add a comment explaining why we have this special case for Apple?

Sure. Apple OpenCL platform (at least its compiler for the CPU) has some kind of bug that makes some kernels impossible to compile. I've figured out that conditions for the bug to show itself are: loop with condition with two variables (e.g. comparing two variables `index < last_index`; if you have `index < 10000` everything is fine) + in this loop you have to write constant or result of function into a buffer (if you're copying value from one buffer to another, e.g. `buf1[idx] = buf2[idx];`, bug does not appear). And that's why some tests work and some don't. 

Examples: [_buf0[i]=42;](https://travis-ci.org/boostorg/compute/jobs/142560405#L2309), [_buf0[i]=ret42();](https://travis-ci.org/boostorg/compute/jobs/142560405#L2365).

The same bug is a problem for vexcl - https://github.com/ddemidov/vexcl/issues/92.